test/k8sT/DatapathConfiguration.go

```diff
@@ -337,7 +375,113 @@ var _ = Describe("K8sDatapathConfig", func() {
+		var (
+			tmpEchoPodPath      string
+			tmpConfigMapDirPath string
+			tmpConfigMapPath    string
+			defaultIface        string
+			err                 error
+		)
+
+		BeforeAll(func() {
+			defaultIface, err = kubectl.GetDefaultIface()
+			Expect(err).Should(BeNil(), "Failed to retrieve default iface")
+
+			// Deploy echoserver on the node which does not run Cilium to test
+			// BPF masquerading. The pod will run in the host netns, so no CNI
+			// is required for the pod on that host.
+			echoPodPath := helpers.ManifestGet(kubectl.BasePath(), "echoserver-hostnetns.yaml")
+			res := kubectl.ExecMiddle("mktemp")
+			res.ExpectSuccess()
+			tmpEchoPodPath = strings.Trim(res.GetStdOut(), "\n")
+			kubectl.ExecMiddle(fmt.Sprintf("sed 's/NODE_WITHOUT_CILIUM/%s/' %s > %s",
+				helpers.GetNodeWithoutCilium(), echoPodPath, tmpEchoPodPath)).ExpectSuccess()
+			kubectl.ApplyDefault(tmpEchoPodPath).ExpectSuccess("Cannot install echoserver application")
+			Expect(kubectl.WaitforPods(helpers.DefaultNamespace, "-l name=echoserver-hostnetns",
+				helpers.HelperTimeout)).Should(BeNil())
+
+			// Setup ip-masq-agent configmap dir
+			res = kubectl.ExecMiddle("mktemp -d")
+			res.ExpectSuccess()
+			tmpConfigMapDirPath = strings.Trim(res.GetStdOut(), "\n")
+			tmpConfigMapPath = filepath.Join(tmpConfigMapDirPath, "config")
+		})
 
+		AfterEach(func() {
+			if tmpConfigMapPath != "" {
+				ns := helpers.GetCiliumNamespace(helpers.GetCurrentIntegration())
+				kubectl.DeleteResource("configmap", fmt.Sprintf("ip-masq-agent --namespace=%s", ns))
+			}
+		})
+
+		AfterAll(func() {
+			if tmpEchoPodPath != "" {
+				kubectl.Delete(tmpEchoPodPath)
+			}
+
+			for _, path := range []string{tmpEchoPodPath, tmpConfigMapPath, tmpConfigMapDirPath} {
+				if path != "" {
+					os.Remove(path)
+				}
+			}
+		})
+
+		testIPMasqAgent := func() {
+			// Check that requests to the echoserver from client pods are masqueraded.
+			nodeIP, err := kubectl.GetNodeIPByLabel(helpers.GetNodeWithoutCilium(), false)
+			Expect(err).Should(BeNil())
+			Expect(testPodHTTPToOutside(kubectl,
+				fmt.Sprintf("http://%s:80", nodeIP), true, false)).Should(BeTrue(),
+				"Connectivity test to http://%s failed", nodeIP)
+
+			// Deploy ip-masq-agent configmap to prevent masquerading to the node IP
+			// which is running the echoserver.
+			kubectl.ExecMiddle(fmt.Sprintf("echo 'nonMasqueradeCIDRs:\n- %s/32' > %s", nodeIP, tmpConfigMapPath)).
+				ExpectSuccess()
+			ns := helpers.GetCiliumNamespace(helpers.GetCurrentIntegration())
+			kubectl.CreateResource("configmap",
+				fmt.Sprintf("ip-masq-agent --from-file=%s --namespace=%s", tmpConfigMapDirPath, ns)).
+				ExpectSuccess("Failed to provision ip-masq-agent configmap")
+
+			// Wait until the ip-masq-agent configmap is mounted into cilium-agent pods,
+			// and the pods have read the new configuration
+			time.Sleep(90 * time.Second)
```

Comment:
:/

I take it that this is done not because of the 1s reaction time we've configured in cilium-agent for changes to ip-masq, but rather because we're relying on the periodic sync of the configmap into the container which may take quite some time.

I would love to see a better approach for the validation, there's three minutes here alone.

Reply:
I think I can poll the ip-masq-agent BPF map entries to detect the update. Added to the follow-up.

Reply:
Even if we poll rather than sleep, I'm guessing that it still may take around this amount of time because we're waiting on the configmap to be updated into the container, no? @aanm could probably tell you how long it takes for a configmap change to become effective inside a container's filesystem.

Reply:
30 - 60 seconds I would say. 0 seconds if the pod is restarted